shared/src/store.rs

```diff
@@ -38,68 +38,38 @@ impl<T: 'static + KeyValueDB> ChainKVStore<T> {
     }
 }
 
-pub struct ChainStoreHeaderIterator<'a, T: ChainStore>
-where
-    T: 'a,
-{
-    store: &'a T,
-    head: Option<Header>,
-}
-
 pub trait ChainStore: Sync + Send {
+    type Batch: StoreBatch;
+    fn new_batch(&self) -> Self::Batch;
+
     fn get_block(&self, block_hash: &H256) -> Option<Block>;
     fn get_header(&self, block_hash: &H256) -> Option<Header>;
     fn get_block_body(&self, block_hash: &H256) -> Option<Vec<Transaction>>;
     fn get_block_proposal_txs_ids(&self, h: &H256) -> Option<Vec<ProposalShortId>>;
     fn get_block_uncles(&self, block_hash: &H256) -> Option<Vec<UncleBlock>>;
     fn get_block_ext(&self, block_hash: &H256) -> Option<BlockExt>;
-    fn insert_block(&self, batch: &mut Batch, b: &Block);
-    fn insert_block_ext(&self, batch: &mut Batch, block_hash: &H256, ext: &BlockExt);
-    fn save_with_batch<F: FnOnce(&mut Batch) -> Result<(), Error>>(
-        &self,
-        f: F,
-    ) -> Result<(), Error>;
-
-    /// Visits block headers backward to genesis.
-    fn headers_iter<'a>(&'a self, head: Header) -> ChainStoreHeaderIterator<'a, Self>
-    where
-        Self: 'a + Sized,
-    {
-        ChainStoreHeaderIterator {
-            store: self,
-            head: Some(head),
-        }
-    }
 }
 
-impl<'a, T: ChainStore> Iterator for ChainStoreHeaderIterator<'a, T> {
-    type Item = Header;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let current_header = self.head.take();
-        self.head = match current_header {
-            Some(ref h) => {
-                if h.number() > 0 {
-                    self.store.get_header(&h.parent_hash())
-                } else {
-                    None
-                }
-            }
-            None => None,
-        };
-        current_header
-    }
+pub trait StoreBatch {
+    fn insert_block(&mut self, block: &Block);
+    fn insert_block_ext(&mut self, block_hash: &H256, ext: &BlockExt);
+    fn insert_tip_header(&mut self, header: &Header);
+
+    fn attach_block(&mut self, block: &Block);
+    fn detach_block(&mut self, block: &Block);
+
+    fn commit(self);
+}
+
+impl<T: KeyValueDB> ChainStore for ChainKVStore<T> {
+    type Batch = DefaultStoreBatch<T::Batch>;
 
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        match self.head {
-            Some(ref h) => (1, Some(h.number() as usize + 1)),
-            None => (0, Some(0)),
+    fn new_batch(&self) -> Self::Batch {
+        DefaultStoreBatch {
+            inner: self.db.batch().expect("new db batch should be ok"),
```

Comment:
If it should be OK, why let the interface returns a result? The two interfaces should be consistent, both use `Result` or neither.

Reply:
Agree, I will change it to use `Result` later

Reply:
Resolved, please review again, thanks